Recover pubkey from signature

TheMhv opened the pull request:
## feature

- recover pubkey from signature

## issue
- fix #3

jaonoctus commented on src/pages/HomeView.vue:
```diff
@@ -27,12 +27,112 @@ import {
+    strToHex(prefix) + 
+    byteArrayToHexString(bech32To8BitArray(data));
+
+  const signingDataParsed = CryptoJS.enc.Hex.parse(signingData);
+  const payHash = CryptoJS.SHA256(signingDataParsed);
```
I don't think you need this CryptoJS lib here. You could use `crypto.subtle.digest('SHA-256', ...)` function instead.

- https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest

jaonoctus reviewed (commented):
I'm happy with the changes. I would just move all those functions to a `lib` or `utils` folder.

And left a small comment about the new dependency which I don't think it's required for this.

And please, leave your lightning address or a invoice for `21,000` sats.

TheMhv commented:
All changes have been made, I moved the functions to the `lib/utils.ts` file and removed the dependency on the `CryptoJS` lib.

Here is the invoice for 21,000 sats:

`lnbc210u1pn4kxl3pp5a8l8aggy0uhjtc4luwtzvt4sktu7gs2p02xm08rxu4rz2g5p0t7sdqqcqzzsxqyz5vqsp5yznmtpj4p3h858ywksnasrvm7rd9zpcv7dkk9268epv2j4vh3shq9qyyssqj26nw6tatx7l6uum79me6nvxw7edq8ym3xgnhydwzmygh0t2vl9ydw7tzpkz65e0ss6fg6edm73vqy24yarnv37tv2gyym5g5elhqscpkj3856`

jaonoctus commented:
@TheMhv the invoice pubkey is wrong for some cases..

look this for example:

https://lightningdecoder.com/lnbc1u1pn4vy88pp5aeat29merqp9ar3gtlzqpsacy65djhy3dtd5wnjdfhe644cxte4sdqqcqzpgxqyz5vqsp54wzqdmnkpc4ukze8xre4wp393aqectlg2tv6nvuh8eg7s3vhm86q9p4gqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqpqysgqx39jzufx3t2nz0kn7694zlusnqx8xxdk6mf596cl2eyg7whl8tvpapgkltll3yh9p3wxmjf5xdtktfk82d7ts57wc3tsc7a0trhfv6cqzmcr6t

it should be `02c4ae20674d7627021639986d75988b5f17c8693ed43b794beeef2384d04e5bf1` but it is `020d62eecaf36bb8be2fc6f90c6c635ca20a674a3632b326a4505d615321ac0df3` with your code.

You must read the bolt11 specs in the issue dev notes. But TLDR, you should not hardcode the recover id as you doing right now in line `secp256k1.ecdsaRecover(new Uint8Array(sigParsed), 1, new Uint8Array(hash), true)`. 

> "UTF-8 bytes, concatenated with the data part (excluding the signature) with 0 bits appended to pad the data to the next byte boundary, with a trailing byte containing the recovery ID (0, 1, 2, or 3)."

jaonoctus commented on src/lib/utils.ts:
```diff
@@ -49,3 +52,93 @@ export function useForm() {
+
+  const hash = await crypto.subtle.digest('SHA-256', hexToArrayBuffer(signingData))
+
+  const signatureValue = signature.value.slice(0, -2)
+  const sigParsed = hexToArrayBuffer(signatureValue)
+
+  const sigPubkey = secp256k1.ecdsaRecover(new Uint8Array(sigParsed), 1, new Uint8Array(hash), true)
```
```suggestion
  const recoveryId = parseInt(signature.value.slice(-2), 16);
  const signatureValue = signature.value.slice(0, -2)
  const sigParsed = hexToArrayBuffer(signatureValue)

  const sigPubkey = secp256k1.ecdsaRecover(new Uint8Array(sigParsed), recoveryId, new Uint8Array(hash), true)
```

jaonoctus commented:
Also squash commits into only 1, please.

jaonoctus reviewed (approved):
tACK

jaonoctus commented:
thank you @TheMhv

here is the proof of payment of the bounty:

https://lnreceipt.jaonoctus.dev/?invoice=lnbc210u1pn4kxl3pp5a8l8aggy0uhjtc4luwtzvt4sktu7gs2p02xm08rxu4rz2g5p0t7sdqqcqzzsxqyz5vqsp5yznmtpj4p3h858ywksnasrvm7rd9zpcv7dkk9268epv2j4vh3shq9qyyssqj26nw6tatx7l6uum79me6nvxw7edq8ym3xgnhydwzmygh0t2vl9ydw7tzpkz65e0ss6fg6edm73vqy24yarnv37tv2gyym5g5elhqscpkj3856&preimage=929d81e5a920141251a8ca5e63dbc2bd3bde0a555e802759d734ef9199861f66
